### PR TITLE
[CPDLP-2105] Add PUT /api/v3/partnerships/ecf/{partnership_id}

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

- Ticket: https://dfedigital.atlassian.net/browse/CPDLP-2105

### Changes proposed in this pull request

* New update partnership endpoint
* New service class `Partnerships::Update`

### CODE REVIEW

**Review by cwrw — COMMENTED** [review]
looking good, just a few comments

File: app/services/partnerships/update.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    def validate_delivery_partner
+      if delivery_partner
+        if !delivery_partner_has_provider_relationship?
+          errors.add(:delivery_partner_id, :no_relationship)
+        elsif partnership.challenged?
+          errors.add(:delivery_partner_id, :partnership_challenged)
+        end
+      else
+        errors.add(:delivery_partner_id, :invalid)
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/services/partnerships/update.rb:49]: Updated 👍

File: app/services/partnerships/update.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    attribute :partnership
+    attribute :delivery_partner_id
+
+    validates :partnership, :delivery_partner_id, presence: true
+    validate :validate_delivery_partner
+
+    def call
+      return false if invalid?
+
+      ::Partnerships::Report.call(
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/services/partnerships/update.rb:18]: Updated 👍

File: app/services/partnerships/update.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    attribute :partnership
+    attribute :delivery_partner_id
+
+    validates :partnership, :delivery_partner_id, presence: true
+    validate :validate_delivery_partner
+
+    def call
+      return false if invalid?
+      return true if partnership.delivery_partner == delivery_partner
```
Comment (cwrw) [inline app/services/partnerships/update.rb:16]: I think we'll need to return the partnership here otherwise it breaks, could you add a test as well?

Comment (mooktakim) [reply app/services/partnerships/update.rb]: Added test

Comment (mooktakim) [reply app/services/partnerships/update.rb]: Added test

**Review by cwrw — APPROVED** [review]
🙌

File: app/services/partnerships/create.rb [hunk 4]
```diff
@@ -100,5 +102,9 @@ def has_previous_cohort_and_matching_lead_provider?
 
       previous_year_lead_provider != lead_provider && school.school_cohorts.find_by(cohort: cohort_record).blank?
     end
+
+    def duplicate_delivery_partner_request?
```
Comment (cwrw) [inline app/services/partnerships/create.rb:106]: thanks for adding this! only comment is we will get duplicate errors with `recruited_by_other_provider` if there is another existing partnership, maybe we can only return this if the other one fails?

**Review by cwrw — APPROVED** [review]
thanks for this, and apologies for the multiple iterations, just one final comment otherwise 👍

File: app/services/partnerships/create.rb (hunk 4 above)
Comment (mooktakim) [reply app/services/partnerships/create.rb:106]: Sure, added a check to see if we already have an error.

**Pull request merged**